### DISCUSSION TITLE
Support for `QuerySource` on `CombinationClause` or similar.

### DISCUSSION THREAD

**Discussion by johnchildren** [diesel-rs/diesel#3249] [Ideas]
It's quite possible there's an obvious way to do this that I'm missing using `Alias`, but I've been struggling with this for a few days and it seems like this is probably not the case yet.

In my diesel code I have a few `UNION` queries where I want to apply a predicate or perform a select, but unfortunately I can't see an obvious way to do this without applying the select or projection to each part of the union. Doing this is tolerable, but leads to a lot of boilerplate when I have a few different filters as I need to apply each filter to each part of the query.

Effectively the kind of query I want to run is:

```sql
SELECT combined.id, combined.name
WHERE combined.age == $1
FROM (
    SELECT foo.id, foo.name, foo.age FROM foo
    UNION
    SELECT bar.id, bar.name, bar.age FROM bar) as combined
```

while currently I am doing:

```sql
    SELECT foo.id, foo.name WHERE foo.age == $1 FROM foo
    UNION
    SELECT bar.id, bar.name WHERE bar.age == $2 FROM bar
```

If I was able to use `CombinationClause` as a `QuerySource` with some kind of aliasing, these queries would be much easier to write from my perspective as I could apply the filters after performing the UNION operation. I think most query planners should be able to push the filters into the sub-queries in my case, though I have not verified that this is the case, so hopefully there wouldn't be a performance problem. 

This would also need some kind of aliasing support to work however, which would require something like the `alias!` macro to give a new name to source and create rust values to represent the columns of the created relation. This would be complicated a bit when the sources of the combination clause are different tables, but I think that would be a necessity to actually implement this feature properly.

Looking at the implementation of `CombinationClause`, I think this might not be easy to do without breaking changes for sqlite as the query becomes `SELECT * FROM ... UNION ...` vs the `... UNION ...` in PostgreSQL and MySQL. I don't know much about sqlite or how to avoid this, but will try and do a bit of research around it.

I have tried to read as many diesel issues as I can around this topic, but there didn't seem to be a specific discussion about it. I'm willing to put in some time to try and figure out what this feature should look like and attempt to implement it, but possibly I'm just doing something silly anyway, so let me know!

**Comment by weiznich** [comment]
First of all: `alias!` and in extension `Alias` currently only supports aliasing tables. It does not support subqueries or column/expression aliasing. This is outlined in the relevant PR. See #2254 for details. We do currently not have any concrete plans to extend aliasing support, at least not before the final 2.0 release is done. Before implementing support for other aliasing types more design work is required on how such a API should look like. [This comment contains](https://github.com/diesel-rs/diesel/pull/2254) at least some information. We are happy to see experimentation and even suggestions here, but please expect that this won't land fast (== likely not in the next few months).

For specific usecases this all is not even required. [The extending diesel guide](https://diesel.rs/guides/extending-diesel.html) outlines how you can easily wrap existing queries and add custom clauses/whatever to the query.

> Looking at the implementation of CombinationClause, I think this might not be easy to do without breaking changes for sqlite as the query becomes SELECT * FROM ... UNION ... vs the ... UNION ... in PostgreSQL and MySQL. I don't know much about sqlite or how to avoid this, but will try and do a bit of research around it.

Can you explain why this would require a breaking change? It's not clear for my based on what you've written so far. Maybe we can fix that before releasing 2.0, if it really requires a breaking change.

**Reply by johnchildren** [reply]
> First of all: `alias!` and in extension `Alias` currently only supports aliasing tables. It does not support subqueries or column/expression aliasing. This is outlined in the relevant PR. See #2254 for details. We do currently not have any concrete plans to extend aliasing support, at least not before the final 2.0 release is done. Before implementing support for other aliasing types more design work is required on how such a API should look like. [This comment contains](https://github.com/diesel-rs/diesel/pull/2254) at least some information.

Thanks, I had missed that thread initially, will try and make sure I have internalized it completely.

> We are happy to see experimentation and even suggestions here, but please expect that this won't land fast (== likely not in the next few months).

Absolutely! Apologies if this thread came off badly.

> For specific usecases this all is not even required. [The extending diesel guide](https://diesel.rs/guides/extending-diesel.html) outlines how you can easily wrap existing queries and add custom clauses/whatever to the query.

Thanks, I will try and see how far I get with this.

> > Looking at the implementation of CombinationClause, I think this might not be easy to do without breaking changes for sqlite as the query becomes SELECT * FROM ... UNION ... vs the ... UNION ... in PostgreSQL and MySQL. I don't know much about sqlite or how to avoid this, but will try and do a bit of research around it.
> 
> Can you explain why this would require a breaking change? It's not clear for my based on what you've written so far. Maybe we can fix that before releasing 2.0, if it really requires a breaking change.

I think I may have been wrong about this, my thinking was that if you wanted to use the `UNION` expression in the `FromClause` for `QuerySource`, for sqlite you would get something like:

```sql
SELECT * FROM (SELECT * FROM (SELECT * FROM foo UNION SELECT * FROM bar))
```
vs for PostgreSQL and MySQL:

```sql
SELECT * FROM (SELECT * FROM foo UNION SELECT * FROM bar) as baz
```

and as a result you would want to change the SQL generated by `CombinationClause` for sqlite to avoid the additional select. But this isn't really necessary as it doesn't change the meaning of the query at all.

I'll try and read up about sqlite more.

edit: my example queries are wrong I think. Will try to fix ASAP.

edit2: I think they are correct now.

**Reply by johnchildren** [reply]
Actually I think you can disregard what I'm saying completely, I think I just misunderstood `ParenthesisWrapper`.

**Comment by johnchildren** [comment]
Having thought about this more and cleared up some of my sqlite misunderstandings, I think what would cover my use-case more is extended alias support for sub-queries and a way to convert `CombinationClause` to an aliased subquery.  This could also be applied to most other query related structs as well.

I think this is discussed a little in the issues linked and won't be easy to implement however. One suggestion might be to allow users to define relations for sub-queries in a similar way to the `alias!` macro and hook into/re-use the alias code.

I could imagine it working something like this:

```rust
table! {
  foos (id) {
    id -> Uuid
    name -> Text
    age -> Int4
  }
}

relation! {
  bars {
     id -> Uuid
     name -> Text
  }
}

// SELECT "foos"."id", "foos"."age" FROM "foos" WHERE "foos"."age" = 24
let subquery = foos::table.filter(foos::age.eq(24)).select((foos::id, foos::name))

// SELECT "bars"."id"
// FROM (SELECT "foos"."id", "foos"."age" FROM "foos" WHERE "foos"."age" = 24) as "bars"
// WHERE "bars"."name" = "Fred"
let query = subquery.as_source(bars::relation).filter(bars::name.eq("Fred")).select((bars::id,))
```

Where `relation!` generates the `bars` consts and an instance of an `Alias` like struct that provides the `QueryFragment` implementation.

For this `.as_source` would need to have a constraint that guarantees the `Expression::SqlType` matches the columns from the `relation` macro. I'm not sure what the ergonomics of this would be though and having to redefine relations like this might be counter to the design approach of diesel.

Joining these relations and using keys after calling `.as_source` makes things a lot more complex though, so I think this could be a somewhat unhelpful suggestion.

**Reply by weiznich** [reply]
It should be definitively possible to experiment with such features as third party crate. It would be helpful to have some concrete implementation ready to experiment a bit with it in terms of usability before deciding which way to go for the "official" solution. That's what we did for the current aliasing support as well, where the first prototype was done by me in some internal project.

**Comment by johnchildren** [comment]
I wrote a bit of an experiment on what this could look like, but I think the way I abuse `Table` here is undesirable. I couldn't see a way to get around it to provide the `AppearsOnTable` trait without doing this but it doesn't really make sense as my subquery isn't a table and isn't guaranteed to have a primary key. Will keep investing and see how this works with joins etc.

https://github.com/johnchildren/diesel-subquery/blob/5fe7809cd1fc06317fa66a3f422e0596f4ba8f4c/src/lib.rs#L194
